0000-interledger-protocol-4/0000-interledger-protocol-4.md

```diff
@@ -0,0 +1,146 @@
+2. Sender sends the ILP Prepare packet to a connector over some communication channel, such as a Websocket. 
+3. Connector checks whether the sender has sufficient bandwidth (balance or credit) to send the amount specified in the packet. If so, the connector reduces the sender's bandwidth by the value of the packet. If not, the connector returns an ILP Reject packet.
+4. Connector uses the destination ILP address from the packet and its local routing tables to determine which next hop to forward the packet to.
+5. Connector modifies the packet amount and expiry to apply their exchange rate and shorten the expiry time.
+6. Connector forwards the packet to the next hop, which may be another connector. All subsequent connectors go through steps 3-6 until the packet reaches the receiver.
+7. Receiver checks the ILP Prepare packet, according to whatever is stipulated by the higher-level protocol (such as checking whether the amount received is above some minimum specified by the sender). 
+8. Receiver returns an ILP Fulfill packet, containing the preimage of the condition in the ILP Prepare packet. The receiver may include additional data, the formatting of which is determined by the higher-level protocol.
+9. Connector checks that the receiver has returned the fulfillment before the expiry in the ILP Prepare packet. If so, the connector returns the same ILP Fulfill packet to the previous connector and credits the receiver's account with the amount in the ILP Prepare packet. If the Prepare has expired, the connector returns an ILP Reject packet to the previous connector. Each connector repeats this step until the Fulfill (or Reject) packet reaches the sender.
+10. Sender checks that the preimage in the ILP Fulfill packet matches the condition in their Prepare packet and may read any data returned by the receiver.
+11. Sender signs an update to their payment channel with the first connector to cover the value of the ILP Prepare (or they may settle less frequently using slower on-ledger transfers or physical deliveries of cash or other goods) and communicates it to the connector.
+12. Connector validates the sender's signed payment channel update (or other proof of settlement) and increases the sender's bandwidth by the amount they paid. If the sender does not pay for the ILP Prepare, the connector may refuse to process further ILP packets until they do or the connector may block that user entirely.
+
+## Specification
+
+### Packet Format
+
+ILPv4 packets SHOULD be encoded using the ASN.1 Octet Encoding Rules.
+
+#### Type-Length Wrapper
+
+All ILP packets are wrapped in the following envelope:
+
+| Field | Type | Description |
+|---|---|---|
+| `type` | UInt8 | ID of the ILP Packet type |
+| `data` | Variable-Length Octet String | Packet contents, prefixed with their length |
```

Comment:
I think the comment in the ASN.1 might contradict the definition (see [here](https://github.com/interledger/rfcs/blob/master/asn1/InterledgerPacket.asn#L57-L58)) but the implementation definitely uses a [length-prefixed field](https://github.com/interledgerjs/ilp-packet/blob/master/index.ts#L32)

Reply:
Hmmm. I believe that is an implementation error. The ASN.1 definition resolves to:

```asn

InterledgerPreparePacket ::= SEQUENCE {
  type UInt8,
  data InterledgerPrepare -- If type == 12 then data is of type InterledgerPrepare
}

InterledgerPrepare ::= SEQUENCE {
  amount UInt64,
  expiresAt Timestamp,
  executionCondition UInt256,
  destination Address,
  data OCTET STRING (SIZE (0..32767))
}
```

In other words it is a `SEQUENCE` as a field within a `SEQUENCE`. There is no length indicator on the field.

Reply:
🤷‍♂️ that might be the kind of implementation error that sticks because changing it would break everything (like [HTTP "Referer"](https://en.wikipedia.org/wiki/HTTP_referer))

Reply:
I think we can change the ASN.1 to reflect the implementation but it will be ugly. We'd lose the binding (in ASN.1) between the `type` and the format.

Something like:

```asn
InterledgerPreparePacket ::= SEQUENCE {
  type UInt8,
  data OCTET STRING
}
```

Then implementors need to know that if `type == 12` then data must be parsed as an `InterledgerPrepare`. It's not pretty.

Reply:
The intended meaning is definitely that there should be a length prefix (otherwise extensibility is broken). I couldn't find a document that describes how `CLASS` is encoded in OER, so I don't know whether the ASN.1 is technically correct. Given that @justmoon tested this in an ASN.1 tool, I assume that `CLASS` does encode with a length prefix.

Reply:
> The intended meaning is definitely that there should be a length prefix (otherwise extensibility is broken).

No, that's why we previously had extensibility in each packet type. An ASN.1 SEQUENCE has no overall length indicator because the sequence of fields is known and their length can be determined independently.

As we've discussed before, this is why you can't send an ILP packet over a transport that doesn't support framing because if you get a bad packet you don't know when the the next good packet starts.

> I couldn't find a document that describes how CLASS is encoded in OER, so I don't know whether the ASN.1 is technically correct.

`CLASS` is not explicitly encoded. It is a way of defining Information Objects which are arrangements of other ASN.1 types. See http://www.oss.com/asn1/resources/asn1-made-simple/advanced-topics.html

In our definition of `PACKET` we specify that a packet must be defined by a UInt8 type and then the packet data.

We then define `PacketSet` which is the acceptable list of types.

This is not valid ASN.1 but basically we define an `InterledgerPreparePacket` as:

```asn
InterledgerPreparePacket ::= SEQUENCE {
  type ::= UInt8 12,
  data ::= SEQUENCE {
    amount UInt64,
    expiresAt Timestamp,
    executionCondition UInt256,
    destination Address,
    data OCTET STRING (SIZE (0..32767))
  }
}
```

Reply:
I used the tool at http://asn1-playground.oss.com/ and it produces the same output as the implementation so either they are both wrong (unlikely) or there is some aspect of OER encoding that I need to understand better (more likely).

```
Encoding to the file 'data.oer' using OER encoding rule...
InterledgerPacket SEQUENCE
  type UInt8 INTEGER [length = 1]
    12
  data InterledgerPrepare SEQUENCE
    amount UInt64 INTEGER [length = 8]
      1000
    expiresAt Timestamp PrintableString [length = 17]
      "20170212201100000"
    executionCondition UInt256 OCTET STRING [length = 32]
      0x0000000000000000000000000000000000 ...
    destination Address IA5String [length = 6]
      "g.test"
    data OCTET STRING [length = 0]
      <no content>
Total encoded length = 67
Encoded successfully in 67 bytes:
0C410000 00000000 03E83230 31373032 31323230 31313030 30303000 00000000
00000000 00000000 00000000 00000000 00000000 00000000 00000006 672E7465
737400

```

`0C` = 12
`41` = 65 (length of data)
`000000 00000003 E8` = 1000 (amount)
`323031 37303231 32323031 31303030 3030` = "20170212201100000" (expiresAt)
`0000 00000000 00000000 00000000 00000000` 
`00000000 00000000 00000000 0000` = 0 (condition as UInt256)
`06` + `67 2E746573 74` = 6 (length of address) + "g.test" (address)
`00` = 0 (length of data)

Reply:
The `InterledgerPacket`'s `data` field is what's called an [open type](http://www.oss.com/asn1/resources/asn1-made-simple/advanced-topics.html). The OER spec says:

> #### 30 Encoding of open type values
>
> NOTE – An open type is an ASN.1 type that can take any abstract value of any ASN.1 type. Each value of an open type consists of:
> a) a contained type; and
> b) a value of the contained type.
>
> The encoding of an open type value shall consist of a **length determinant (see 8.6)** followed by a series of octets, which are the encoding of the value of the contained type.

*(emphasis added)*

So both OSS' and our implementations are both correct in adding a length determinant.

Reply:
I realized when implementing this, the reason for having the length and it's important for implementors to be aware of this if they wish to be forward-compatible (although this is less of an issue in ILPv4 because there is not case when you would forward a packet with a type you don't know.)

Since this is an Open Type, a decoder may encounter a `type` value they don't recognize (e.g. 15) and therefor don't know how to decode the associated `data`. There may be situations where this is not a critical error and so the decoder can simply consume the correct number of bytes and continue decoding the rest of the message (and possibly forward these un-parsable bytes to the next entity).